### PR TITLE
Libscion enhancements

### PR DESCRIPTION
**Pull request by aznair**
- get_path utility function
- change type of buf argument to void \* for all functions
- simplify signature of reverse_path function
- add get_total_header_len, get_total_ext_len functions
- move uthash to libscion
  <a href='#crh-start'></a><a href='#crh-data-%7B%22processed%22%3A%20%5B%22https%3A//github.com/netsec-ethz/scion/pull/700%23issuecomment-210360324%22%2C%20%22https%3A//github.com/netsec-ethz/scion/pull/700%23discussion_r59862937%22%2C%20%22https%3A//github.com/netsec-ethz/scion/pull/700%23discussion_r59863030%22%2C%20%22https%3A//github.com/netsec-ethz/scion/pull/700%23issuecomment-210436110%22%2C%20%22https%3A//github.com/netsec-ethz/scion/pull/700%23discussion_r59863068%22%2C%20%22https%3A//github.com/netsec-ethz/scion/pull/700%23discussion_r60365130%22%2C%20%22https%3A//github.com/netsec-ethz/scion/pull/700%23discussion_r60365388%22%2C%20%22https%3A//github.com/netsec-ethz/scion/pull/700%23discussion_r60365984%22%2C%20%22https%3A//github.com/netsec-ethz/scion/pull/700%23discussion_r60366002%22%2C%20%22https%3A//github.com/netsec-ethz/scion/pull/700%23discussion_r60366314%22%2C%20%22https%3A//github.com/netsec-ethz/scion/pull/700%23discussion_r60366720%22%2C%20%22https%3A//github.com/netsec-ethz/scion/pull/700%23discussion_r60366969%22%2C%20%22https%3A//github.com/netsec-ethz/scion/pull/700%23discussion_r60367567%22%2C%20%22https%3A//github.com/netsec-ethz/scion/pull/700%23discussion_r60367699%22%2C%20%22https%3A//github.com/netsec-ethz/scion/pull/700%23issuecomment-215020943%22%2C%20%22https%3A//github.com/netsec-ethz/scion/pull/700%23discussion_r61226569%22%2C%20%22https%3A//github.com/netsec-ethz/scion/pull/700%23discussion_r61226757%22%2C%20%22https%3A//github.com/netsec-ethz/scion/pull/700%23discussion_r61226909%22%2C%20%22https%3A//github.com/netsec-ethz/scion/pull/700%23discussion_r61226978%22%2C%20%22https%3A//github.com/netsec-ethz/scion/pull/700%23discussion_r61227058%22%2C%20%22https%3A//github.com/netsec-ethz/scion/pull/700%23discussion_r61227087%22%2C%20%22https%3A//github.com/netsec-ethz/scion/pull/700%23discussion_r61227372%22%2C%20%22https%3A//github.com/netsec-ethz/scion/pull/700%23discussion_r61227422%22%2C%20%22https%3A//github.com/netsec-ethz/scion/pull/700%23issuecomment-215025854%22%5D%2C%20%22comments%22%3A%20%7B%22General%20Comment%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/netsec-ethz/scion/pull/700%23issuecomment-210360324%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22These%20are%20some%20changes%20I%20needed%20while%20working%20on%20HSR%2C%20would%20also%20appreciate%20suggestions%20on%20features%20to%20add%20to%20libscion%5Cr%5Cn%5Cr%5Cn%40pszalach%20%40kormat%20%22%2C%20%22created_at%22%3A%20%222016-04-15T08%3A35%3A11Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/801826%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/aznair%22%7D%7D%2C%20%7B%22body%22%3A%20%22otherwise%20lgtm%22%2C%20%22created_at%22%3A%20%222016-04-15T11%3A50%3A55Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/8159928%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/pszalach%22%7D%7D%2C%20%7B%22body%22%3A%20%22I%20thought%20there%20was%20more%20stuff%20I%20wanted%20to%20add%20here%20but%20can%27t%20remember%20what%20it%20was%2C%20so%20this%20will%20be%20it%20for%20now.%22%2C%20%22created_at%22%3A%20%222016-04-27T09%3A11%3A35Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/801826%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/aznair%22%7D%7D%2C%20%7B%22body%22%3A%20%22LGTM%22%2C%20%22created_at%22%3A%20%222016-04-27T09%3A28%3A59Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/6919822%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/kormat%22%7D%7D%5D%2C%20%22title%22%3A%20%22General%20Comment%22%7D%2C%20%22Pull%20b97243270acf0e28ef9b78758f2f4d874f0a8657%20lib/libscion/packet.c%2077%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/netsec-ethz/scion/pull/700%23discussion_r60366314%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22It%20seems%20it%20would%20be%20easier%20to%20just%20do%3A%5Cr%5Cn%60%60%60%5Cr%5Cnntohs%28sch-%3Eheader_len%29%20-%20sizeof%28SCIONCommonHeader%29%20-%20padded_addr_len%28buf%29%3B%5Cr%5Cn%60%60%60%22%2C%20%22created_at%22%3A%20%222016-04-20T08%3A12%3A39Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/6919822%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/kormat%22%7D%7D%2C%20%7B%22body%22%3A%20%22%3A%2B1%3A%20%22%2C%20%22created_at%22%3A%20%222016-04-27T09%3A18%3A16Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/6919822%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/kormat%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20lib/libscion/packet.c%3AL91-141%22%7D%2C%20%22Pull%20a64f9f747c756b93429809754758124f36609921%20lib/libscion/packet.c%2085%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/netsec-ethz/scion/pull/700%23discussion_r59863030%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22maybe%20change%20the%20name%2C%20as%20it%20returns%20hdr_len%20%2B%20ext_len%22%2C%20%22created_at%22%3A%20%222016-04-15T11%3A50%3A38Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/8159928%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/pszalach%22%7D%7D%2C%20%7B%22body%22%3A%20%22%3A%2B1%3A%20%22%2C%20%22created_at%22%3A%20%222016-04-27T09%3A15%3A59Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/6919822%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/kormat%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20lib/libscion/packet.c%3AL91-141%22%7D%2C%20%22Pull%20b97243270acf0e28ef9b78758f2f4d874f0a8657%20lib/libscion/extensions.c%2022%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/netsec-ethz/scion/pull/700%23discussion_r60365388%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22Why%20is%20this%20check%20in%20so%20many%20functions%3F%22%2C%20%22created_at%22%3A%20%222016-04-20T08%3A04%3A32Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/6919822%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/kormat%22%7D%7D%2C%20%7B%22body%22%3A%20%22Paranoia%22%2C%20%22created_at%22%3A%20%222016-04-20T08%3A10%3A05Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/801826%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/aznair%22%7D%7D%2C%20%7B%22body%22%3A%20%22Unless%20you%27re%20checking%20for%20this%20error%20in%20every%20single%20place%20that%20calls%20every%20one%20of%20these%20functions%2C%20you%20haven%27t%20gained%20anything.%20Also%20if%20a%20null%20pointer%20is%20being%20passed%20around%2C%20then%20you%20already%20have%20a%20serious%20problem%2C%20so%20crashing%20is%20probably%20the%20right%20thing%20to%20do.%22%2C%20%22created_at%22%3A%20%222016-04-20T08%3A18%3A07Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/6919822%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/kormat%22%7D%7D%2C%20%7B%22body%22%3A%20%22E.g.%20if%20you%20access%20a%20null%20pointer%2C%20you%27re%20guaranteed%20to%20get%20a%20crash.%20If%20a%20caller%20gets%20a%20-1%2C%20the%20results%20are%20unpredictable.%22%2C%20%22created_at%22%3A%20%222016-04-20T08%3A24%3A14Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/6919822%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/kormat%22%7D%7D%2C%20%7B%22body%22%3A%20%22%3A%2B1%3A%20%22%2C%20%22created_at%22%3A%20%222016-04-27T09%3A17%3A29Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/6919822%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/kormat%22%7D%7D%2C%20%7B%22body%22%3A%20%22I%20think%20it%20would%20be%20nice%20to%20log%20this%20in%20case%20it%20happens%20%28as%20it%20means%20that%20something%20went%20wrong%29.%20About%20letting%20it%20crash%20though%2C%20I%20am%20not%20sure%2C%20we%20already%20have%20enough%20number%20of%20crashes%20in%20the%20system%20I%20believe%2C%20we%20should%20not%20add%20more%20%3B-%29%20%22%2C%20%22created_at%22%3A%20%222016-04-27T09%3A18%3A01Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/7704593%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/ercanucan%22%7D%7D%2C%20%7B%22body%22%3A%20%22After%20reading%20Stephens%27%20comments%20and%20thinking%20about%20it%20some%20more%2C%20the%20NULL%20arguments%20would%20only%20happen%20if%20some%20sort%20of%20memory%20allocation%20failed%20or%20I%20screwed%20up%20something%20in%20the%20code.%20Either%20way%20crashing%20seems%20a%20good%20way%20to%20go.%22%2C%20%22created_at%22%3A%20%222016-04-27T09%3A20%3A37Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/801826%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/aznair%22%7D%7D%2C%20%7B%22body%22%3A%20%22If%20you%20get%20a%20null%20pointer%2C%20something%20has%20gone%20seriously%20wrong%20in%20the%20code%2C%20and%20the%20state%20can%20probably%20no%20longer%20be%20trusted.%20You%20could%2C%20in%20theory%2C%20add%20massive%20amounts%20of%20error%20handling%20to%20try%20and%20recover%20from%20this%2C%20but%20it%20would%20greatly%20complicate%20the%20code%2C%20and%20it%27s%20not%20clear%20that%20it%20would%20actually%20help.%20Adding%20a%20log%20message%20everywhere%20a%20pointer%20might%20turn%20out%20to%20be%20null%20is%20a%20huge%20amount%20of%20logging%20statements%2C%20which%20hurts%20readability%2C%20for%20not%20much%20gain.%20To%20understand%20what%20happened%2C%20you%20need%20a%20backtrace%2C%20and%20probably%20a%20debugger%20attached.%22%2C%20%22created_at%22%3A%20%222016-04-27T09%3A20%3A57Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/6919822%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/kormat%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20lib/libscion/extensions.c%3AL37-64%22%7D%2C%20%22Pull%20b97243270acf0e28ef9b78758f2f4d874f0a8657%20lib/libscion/address.c%207%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/netsec-ethz/scion/pull/700%23discussion_r60365130%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22This%20is%20illegal%3A%20http%3A//stackoverflow.com/a/3524270%5Cr%5Cn%5Cr%5CnWhy%20the%20change%20to%20%60void%20%2A%60%3F%22%2C%20%22created_at%22%3A%20%222016-04-20T08%3A02%3A29Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/6919822%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/kormat%22%7D%7D%2C%20%7B%22body%22%3A%20%22The%20idea%20was%20to%20avoid%20the%20need%20for%20casting%20to%20uint8_t%20%2A%20before%20calling%20all%20these%20functions%20in%20cases%20where%20the%20caller%20has%20e.g.%20a%20SCIONCommonHeader%20pointer%20but%20not%20a%20uint8_t%20pointer%20which%20happens%20fairly%20often%20in%20HSR%22%2C%20%22created_at%22%3A%20%222016-04-20T08%3A09%3A56Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/801826%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/aznair%22%7D%7D%2C%20%7B%22body%22%3A%20%22Well%20the%20consequence%20is%20that%20you%20need%20to%20do%20a%20cast%20to%20uint8_t%20in%20every%20function%2C%20which%20doesn%27t%20seem%20like%20a%20win.%22%2C%20%22created_at%22%3A%20%222016-04-20T08%3A16%3A10Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/6919822%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/kormat%22%7D%7D%2C%20%7B%22body%22%3A%20%22Also%20the%20casting%20everywhere%20mostly%20goes%20away%20with%20my%20proposed%20change%20to%20libscion%20%28i.e.%20a%20meta-structure%20that%20pointers%20to%20all%20the%20various%20parts%20of%20the%20packet%29.%22%2C%20%22created_at%22%3A%20%222016-04-20T08%3A23%3A06Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/6919822%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/kormat%22%7D%7D%2C%20%7B%22body%22%3A%20%22%3A%2B1%3A%20%22%2C%20%22created_at%22%3A%20%222016-04-27T09%3A16%3A59Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/6919822%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/kormat%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20lib/libscion/address.c%3AL15-22%22%7D%2C%20%22Pull%20a64f9f747c756b93429809754758124f36609921%20lib/libscion/packet.c%2077%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/netsec-ethz/scion/pull/700%23discussion_r59862937%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22does%20it%20really%20return%20path%20len%3F%22%2C%20%22created_at%22%3A%20%222016-04-15T11%3A49%3A38Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/8159928%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/pszalach%22%7D%7D%2C%20%7B%22body%22%3A%20%22yeah%20that%20should%20probably%20be%20header_len%20not%20total_len%22%2C%20%22created_at%22%3A%20%222016-04-15T11%3A51%3A01Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/801826%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/aznair%22%7D%7D%2C%20%7B%22body%22%3A%20%22%3A%2B1%3A%20%22%2C%20%22created_at%22%3A%20%222016-04-27T09%3A14%3A55Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/6919822%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/kormat%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20lib/libscion/packet.c%3AL91-141%22%7D%7D%7D'></a>
  <a href='https://www.codereviewhub.com/'></a>
- [ ] <a href='#crh-comment-General Comment'></a> &nbsp;<b><a href='https://github.com/netsec-ethz/scion/pull/700#issuecomment-210360324'>General Comment</a></b>
- <a href='https://github.com/aznair'></a> These are some changes I needed while working on HSR, would also appreciate suggestions on features to add to libscion
  @pszalach @kormat
- <a href='https://github.com/pszalach'></a> otherwise lgtm
- <a href='https://github.com/aznair'></a> I thought there was more stuff I wanted to add here but can't remember what it was, so this will be it for now.
- <a href='https://github.com/kormat'></a> LGTM
- [ ] <a href='#crh-comment-Pull b97243270acf0e28ef9b78758f2f4d874f0a8657 lib/libscion/extensions.c 22'></a> &nbsp;<b><a href='https://github.com/netsec-ethz/scion/pull/700#discussion_r60365388'>File: lib/libscion/extensions.c:L37-64</a></b>
- <a href='https://github.com/kormat'></a> Why is this check in so many functions?
- <a href='https://github.com/aznair'></a> Paranoia
- <a href='https://github.com/kormat'></a> Unless you're checking for this error in every single place that calls every one of these functions, you haven't gained anything. Also if a null pointer is being passed around, then you already have a serious problem, so crashing is probably the right thing to do.
- <a href='https://github.com/kormat'></a> E.g. if you access a null pointer, you're guaranteed to get a crash. If a caller gets a -1, the results are unpredictable.
- <a href='https://github.com/kormat'></a> :+1:
- <a href='https://github.com/ercanucan'></a> I think it would be nice to log this in case it happens (as it means that something went wrong). About letting it crash though, I am not sure, we already have enough number of crashes in the system I believe, we should not add more ;-)
- <a href='https://github.com/aznair'></a> After reading Stephens' comments and thinking about it some more, the NULL arguments would only happen if some sort of memory allocation failed or I screwed up something in the code. Either way crashing seems a good way to go.
- <a href='https://github.com/kormat'></a> If you get a null pointer, something has gone seriously wrong in the code, and the state can probably no longer be trusted. You could, in theory, add massive amounts of error handling to try and recover from this, but it would greatly complicate the code, and it's not clear that it would actually help. Adding a log message everywhere a pointer might turn out to be null is a huge amount of logging statements, which hurts readability, for not much gain. To understand what happened, you need a backtrace, and probably a debugger attached.
- [x] <a href='#crh-comment-Pull a64f9f747c756b93429809754758124f36609921 lib/libscion/packet.c 77'></a> &nbsp;<b><a href='https://github.com/netsec-ethz/scion/pull/700#discussion_r59862937'>File: lib/libscion/packet.c:L91-141</a></b>
- <a href='https://github.com/pszalach'></a> does it really return path len?
- <a href='https://github.com/aznair'></a> yeah that should probably be header_len not total_len
- [x] <a href='#crh-comment-Pull a64f9f747c756b93429809754758124f36609921 lib/libscion/packet.c 85'></a> &nbsp;<b><a href='https://github.com/netsec-ethz/scion/pull/700#discussion_r59863030'>File: lib/libscion/packet.c:L91-141</a></b>
- <a href='https://github.com/pszalach'></a> maybe change the name, as it returns hdr_len + ext_len
- [x] <a href='#crh-comment-Pull b97243270acf0e28ef9b78758f2f4d874f0a8657 lib/libscion/address.c 7'></a> &nbsp;<b><a href='https://github.com/netsec-ethz/scion/pull/700#discussion_r60365130'>File: lib/libscion/address.c:L15-22</a></b>
- <a href='https://github.com/kormat'></a> This is illegal: http://stackoverflow.com/a/3524270
  Why the change to `void *`?
- <a href='https://github.com/aznair'></a> The idea was to avoid the need for casting to uint8_t \* before calling all these functions in cases where the caller has e.g. a SCIONCommonHeader pointer but not a uint8_t pointer which happens fairly often in HSR
- <a href='https://github.com/kormat'></a> Well the consequence is that you need to do a cast to uint8_t in every function, which doesn't seem like a win.
- <a href='https://github.com/kormat'></a> Also the casting everywhere mostly goes away with my proposed change to libscion (i.e. a meta-structure that pointers to all the various parts of the packet).
- [x] <a href='#crh-comment-Pull b97243270acf0e28ef9b78758f2f4d874f0a8657 lib/libscion/packet.c 77'></a> &nbsp;<b><a href='https://github.com/netsec-ethz/scion/pull/700#discussion_r60366314'>File: lib/libscion/packet.c:L91-141</a></b>
- <a href='https://github.com/kormat'></a> It seems it would be easier to just do:

```
ntohs(sch->header_len) - sizeof(SCIONCommonHeader) - padded_addr_len(buf);
```

<a href='https://www.codereviewhub.com/netsec-ethz/scion/pull/700?mark_as_completed=1'></a>&nbsp;<a href='https://www.codereviewhub.com/netsec-ethz/scion/pull/700?approve=1'></a>&nbsp;<a href='https://github.com/netsec-ethz/scion/pull/700'></a>
<a href='#crh-end'></a>

### CODE REVIEW

File: lib/libscion/packet.c [hunk 1]
```diff
@@ -79,10 +91,50 @@ void set_path(uint8_t *buf, uint8_t *path, int len)
...
+
+/*
+ * Get length of SCION path
+ * buf: Pointer to start of SCION packet
+ * return value: Length of SCION path in packet
+ */
+int get_path_len(void *buf)
+{
+    SCIONCommonHeader *sch = (SCIONCommonHeader *)buf;
+    return ntohs(sch->total_len) - (get_path(buf) - (uint8_t *)buf);
```
Comment (pszal) [inline lib/libscion/packet.c]: does it really return path len?

File: lib/libscion/packet.c [hunk 2]
```diff
@@ -79,10 +91,50 @@ void set_path(uint8_t *buf, uint8_t *path, int len)
...
+}
+
+/*
+ * Get total length of headers including extensions
+ * buf: Pointer to start of SCION packet
+ * return value: Length of combined SCION headers, -1 on error
+ */
+int get_total_header_len(void *buf)
```
Comment (pszal) [inline lib/libscion/packet.c]: maybe change the name, as it returns hdr_len + ext_len

File: lib/libscion/packet.c (hunk 1 above)
Comment (aznair) [inline lib/libscion/packet.c]: yeah that should probably be header_len not total_len

File: lib/libscion/address.c [hunk 3]
```diff
@@ -15,7 +15,7 @@ const int ADDR_LENS[] = {0, 4, 16, 2};
  * buf: Pointer to start of SCION packet
  * return value: src ISD_AS value, 0 on error
  */
-uint32_t get_src_isd_as(uint8_t *buf)
+uint32_t get_src_isd_as(void *buf)
 {
     return ntohl(*(uint32_t *)(buf + sizeof(SCIONCommonHeader)));
```
Comment (kormat) [inline lib/libscion/address.c]: This is illegal: http://stackoverflow.com/a/3524270

Why the change to `void *`?

File: lib/libscion/extensions.c [hunk 4]
```diff
@@ -37,3 +37,27 @@ uint8_t * find_extension(uint8_t *buf, uint8_t ext_class, uint8_t ext_type)
...
+
+/*
+ * Get total length of all extension headers
+ * buf: Pointer to start of SCION packet
+ * return value: Total length of all extension headers
+ */
+int get_total_ext_len(void *buf)
+{
+    if (!buf)
+        return -1;
```
Comment (kormat) [inline lib/libscion/extensions.c]: Why is this check in so many functions?

File: lib/libscion/address.c (hunk 3 above)
Comment (aznair) [inline lib/libscion/address.c]: The idea was to avoid the need for casting to uint8_t \* before calling all these functions in cases where the caller has e.g. a SCIONCommonHeader pointer but not a uint8_t pointer which happens fairly often in HSR

File: lib/libscion/extensions.c (hunk 4 above)
Comment (aznair) [inline lib/libscion/extensions.c]: Paranoia

File: lib/libscion/packet.c [hunk 5]
```diff
@@ -79,10 +91,50 @@ void set_path(uint8_t *buf, uint8_t *path, int len)
...
+
+/*
+ * Get length of SCION path
+ * buf: Pointer to start of SCION packet
+ * return value: Length of SCION path in packet
+ */
+int get_path_len(void *buf)
+{
+    SCIONCommonHeader *sch = (SCIONCommonHeader *)buf;
+    return ntohs(sch->header_len) - (get_path(buf) - (uint8_t *)buf);
```
Comment (kormat) [inline lib/libscion/packet.c]: It seems it would be easier to just do:

```
ntohs(sch->header_len) - sizeof(SCIONCommonHeader) - padded_addr_len(buf);
```

File: lib/libscion/address.c (hunk 3 above)
Comment (kormat) [inline lib/libscion/address.c]: Well the consequence is that you need to do a cast to uint8_t in every function, which doesn't seem like a win.

File: lib/libscion/extensions.c (hunk 4 above)
Comment (kormat) [inline lib/libscion/extensions.c]: Unless you're checking for this error in every single place that calls every one of these functions, you haven't gained anything. Also if a null pointer is being passed around, then you already have a serious problem, so crashing is probably the right thing to do.

File: lib/libscion/address.c (hunk 3 above)
Comment (kormat) [inline lib/libscion/address.c]: Also the casting everywhere mostly goes away with my proposed change to libscion (i.e. a meta-structure that pointers to all the various parts of the packet).

File: lib/libscion/extensions.c (hunk 4 above)
Comment (kormat) [inline lib/libscion/extensions.c]: E.g. if you access a null pointer, you're guaranteed to get a crash. If a caller gets a -1, the results are unpredictable.

File: lib/libscion/packet.c (hunk 1 above)
Comment (kormat) [inline lib/libscion/packet.c]: :+1:

File: lib/libscion/packet.c (hunk 2 above)
Comment (kormat) [inline lib/libscion/packet.c]: :+1:

File: lib/libscion/address.c (hunk 3 above)
Comment (kormat) [inline lib/libscion/address.c]: :+1:

File: lib/libscion/extensions.c (hunk 4 above)
Comment (kormat) [inline lib/libscion/extensions.c]: :+1:

Comment (ercanucan) [inline lib/libscion/extensions.c]: I think it would be nice to log this in case it happens (as it means that something went wrong). About letting it crash though, I am not sure, we already have enough number of crashes in the system I believe, we should not add more ;-)

File: lib/libscion/packet.c (hunk 5 above)
Comment (kormat) [inline lib/libscion/packet.c]: :+1:

File: lib/libscion/extensions.c (hunk 4 above)
Comment (aznair) [inline lib/libscion/extensions.c]: After reading Stephens' comments and thinking about it some more, the NULL arguments would only happen if some sort of memory allocation failed or I screwed up something in the code. Either way crashing seems a good way to go.

Comment (kormat) [inline lib/libscion/extensions.c]: If you get a null pointer, something has gone seriously wrong in the code, and the state can probably no longer be trusted. You could, in theory, add massive amounts of error handling to try and recover from this, but it would greatly complicate the code, and it's not clear that it would actually help. Adding a log message everywhere a pointer might turn out to be null is a huge amount of logging statements, which hurts readability, for not much gain. To understand what happened, you need a backtrace, and probably a debugger attached.
